mautic/mautic
# Aggregate stat service

| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "features" 
| Bug fix?                               | no
| New feature?                           | yes
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | yes
| Related user documentation PR URL      | na
| Related developer documentation PR URL | 
| Issue(s) addressed                     | 

This adds the service to use a new event to collect stats to allow 3rd party to manage and inject stats. The email graph's TEMPLATE graph has been rewritten to use it. Note that a lot of the statistic collection for the email was refactored into individual classes an are not necessarily part of the example for how to use the service. 

(Note this PR is a blocker to Acquia contributing a number of PRs as code changed depends on this code).

https://github.com/mautic/mautic/pull/9757/files#diff-a9a319bd79a694d0b73f80b6ceb80b383601bd77b0d943890d35eb5275f1d6b8R65 demonstrates how to dispatch a request to collect a specific set of stats (by hour) and formats it for whatever its purpose is. In this case, it's the data for chart.js.

https://github.com/mautic/mautic/pull/9757/files#diff-20b8dee755b891bca4995d9bfe8f2d065b0ec8577261fda99bc1fec627eea70dR44 listens to the event dispatched by the collector service. 

https://github.com/mautic/mautic/pull/9757/files#diff-a9a319bd79a694d0b73f80b6ceb80b383601bd77b0d943890d35eb5275f1d6b8R145 injects the stats into the StatCollection.

To test this PR for the email graph, 
1. create a template email. 
2. create a campaign to send that email to some contacts. 
3. Open some of the emails, click on the links in some of the emails, and unsubscribe from one of the emails. 
4. View the email details page to see if the stats are legit.
